### DISCUSSION TITLE
How does on.paths really work?

### DISCUSSION THREAD

**Discussion by pbrisbin** [community/community#25767] [Actions]
<p>We make heavy use of</p>
<pre><code class="lang-yaml">on:
  push:
    paths:
      - "something/**"
</code></pre>
<p>For scoping Workflows related to some sub-component of our monorepository to only run when code within that component has changed.</p>
<p>We find <em>very frequently</em> the Workflows run when we’re not expecting them to. Is there any documentation on <em>exactly</em> how the “changed paths” list that is matched to this glob is computed?</p>
<p>From a user’s perspective, I’d expect it to match the list of changed files in the PR diff view at the time of push. Technically speaking, I’d expect it to be:</p>
<pre><code class="lang-auto">mb=$(git merge-base "$base_branch" HEAD)
git diff --name-only "$mb"
</code></pre>
<p>But we see these scoped Workflows run when there are no matched paths by those approaches.</p>
<p>Anecdotally, it seems to be related to syncing from the base branch. When we merge in <code>main</code> or rebase + force-push is when the Workflows seem to run unnecessarily.</p>

**Comment by jsoref** [comment, marked as answer]
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#git-diff-comparisons" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#git-diff-comparisons" target="_blank" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#git-diff-comparisons" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/workflow-syntax-for-github-actions</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<blockquote>
<p><strong>Note:</strong> If you push more than 1,000 commits, or if GitHub does not generate the diff due to a timeout (diffs that are too large diffs), the workflow will always run.</p>
</blockquote>
<p>(Amusingly, this isn’t an edge I’ve hit much. I do regularly hit the <a href="https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/limits-for-viewing-content-and-diffs-in-a-repository#commit-listings-limits" rel="noopener nofollow ugc">250 commit limit</a>.)</p>
<blockquote>
<p>GitHub generates the list of changed files using two-dot diffs for pushes and three-dot diffs for pull requests:</p>
</blockquote>
<blockquote>
<ul>
<li>
<strong>Pull requests:</strong> Three-dot diffs are a comparison between the most recent version of the topic branch and the commit where the topic branch was last synced with the base branch.</li>
<li>
<strong>Pushes to existing branches:</strong> A two-dot diff compares the head and base SHAs directly with each other.</li>
<li>
<strong>Pushes to new branches:</strong> A two-dot diff against the parent of the ancestor of the deepest commit pushed.</li>
</ul>
</blockquote>

**Comment by pbrisbin** [comment]
<p>Thanks for finding that! Next time we find unexpected runs, I’ll look to see if this explains it.</p>

**Comment by alanwilter** [comment]
<p>I’m still lost, I have a similar problem. All I want is to run <code>codeql-analysis.yml</code> when it matters:</p>
<p>I have my <code>master</code> branch.</p>
<p>I created a <code>feature</code> branch and opened a PR to <code>master</code> branch.</p>
<p>During a cycle, I may change python files in <code>folderA</code>. In another push I may only change <code>docs</code>. And so on. Right now, whatever I try to set up in my eg. <code>.github/workflows/codeql-analysis.yml</code>, it will run anyway after any push I do to branch <code>feature</code> and I don’t want this happen.</p>
<p>I don’t want all my workflow checks to run unless I change something in two specific folders, say <code>folderA</code>  and <code>folderB</code>, where I have my <code>*.py</code> files.</p>
<p>So, whenever I change something in <code>docs</code> or even <code>.github/workflows/*.yml</code> but not changing anything about <code>*.py</code> in one of the mentioned folders, I <strong>don’t want</strong> the workflow to run whenever I do a push!</p>
<p>I have this for my <code>.github/workflows/codeql-analysis.yml</code>:</p>
<pre><code class="lang-auto">name: "CodeQL"

on:
  push:
    branches: [master] 

  pull_request:
    branches: [master]
    paths:
      - "folderA/**"
      - "folderB/**"
  schedule:
    - cron: "15 0 * * *"
...
</code></pre>
<p>So how can I tweak my <code>codeql-analysis.yml</code> to run it only when it matters?</p>

**Comment by alanwilter** [comment]
<p>Well, by the hard way, as usual, I got to understand what <a class="mention" href="/jsoref">@jsoref</a> pointed out.</p>
<p>So my <code>yml</code> is “ok-ish”. The problem is since it’s a PR and the diff is from latest to base, once I commit anything from the <code>folderA/b</code>, boom, whatever newer push I do will trigger the workflow.</p>
<p>So, it’s ok when I do a PR just to fix docs etc., but if I need to add a <code>*.py</code> change, it will call the workflow as expected… and for any later other push I do, which I wished not.</p>
<p>What I was hoping is that the diff comparison were done per sequential PUSH.</p>
<p>Probably there’s a solution for what I wanted. Any hints?</p>

**Reply by Achllle** [reply]
It would be nice to have this as a flag you can set, I think a lot of people run into this.

**Reply by jsoref** [reply]
As noted, you can implement it.

The thing is that, on average, if a CI author for a project has added this, they _probably_ usually want to validate the whole thing.

Imagine that you have a PR which at tip breaks two things. One is in a thing matched by `on.paths`, the other isn't. You then fix the thing that isn't matched by `on.pull_request.paths` in a new commit and push to your branch which is tracked by your PR. The `on.pull_request` as defined today will run and detect you have still broken the thing that it cares about. If it did what you are asking for, it wouldn't, and then the poor repository owner might merge something that's broken, and that'd be bad.

If you want this behavior, you could switch to `on.push` instead of `on.pull_request` since it will _probably_ do what you want, in that it will compare your old commit to your current commit and thus not pay attention to your accumulated and unchanged commits.

**Comment by jsoref** [comment]
<p>There are some github actions that probably try to help here. (I’m fairly certain I’ve seen some that will help with this.)</p>
<p>A quick search shows:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/tj-actions/changed-files">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/tj-actions/changed-files" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181093175-06259a75-1e31-4b40-8ac1-9d0e4c5ba18e.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/0/805e354d46632f1c36bbc38802af7abb1632bb12_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/0/805e354d46632f1c36bbc38802af7abb1632bb12_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/0/805e354d46632f1c36bbc38802af7abb1632bb12.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/0/805e354d46632f1c36bbc38802af7abb1632bb12_2_10x10.png"></div>

<h3><a href="https://github.com/tj-actions/changed-files" target="_blank" rel="noopener nofollow ugc">GitHub - tj-actions/changed-files: Github action to retrieve all (added,...</a></h3>

  <p>:octocat: Github action to retrieve all (added, copied, modified, deleted, renamed, type changed, unmerged, unknown) files with support for push and pull request events. - GitHub - tj-actions/chang...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
To make it work would require a bit of magic, since you’d have to fish for the previous head value.</p>
<p>I’m fairly certain it’s floating around, in fact, check-spelling has some magic for it:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/unknown-words.sh#L963-L980">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/unknown-words.sh#L963-L980" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/unknown-words.sh#L963-L980" target="_blank" rel="noopener nofollow ugc">check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/unknown-words.sh#L963-L980</a></h4>



  <pre class="onebox">    <code class="lang-sh">
      <ol class="start lines" start="963" style="counter-reset: li-counter 962 ;">
          <li>if to_boolean "$INPUT_ONLY_CHECK_CHANGED_FILES"; then</li>
          <li>  COMPARE=$(cat "$GITHUB_EVENT_PATH" | jq -r '.compare // empty' 2&gt;/dev/null)</li>
          <li>  if [ -n "$COMPARE" ]; then</li>
          <li>    BEFORE=$(echo "$COMPARE" | perl -ne 'if (m{/compare/(.*)\.\.\.}) { print $1; } elsif (m{/commit/([0-9a-f]+)$}) { print "$1^"; };')</li>
          <li>    BEFORE=$(curl -s \</li>
          <li>      -H "$AUTHORIZATION_HEADER" \</li>
          <li>      "$GITHUB_API_URL/repos/$GITHUB_REPOSITORY/commits/$BEFORE" | jq -r '.sha // empty')</li>
          <li>  elif [ -n "$GITHUB_BASE_REF" ]; then</li>
          <li>    BEFORE=$GITHUB_BASE_REF</li>
          <li>  fi</li>
          <li>fi</li>
          <li>if [ -n "$BEFORE" ]; then</li>
          <li>  echo "Only checking files changed from $BEFORE" &gt;&amp;2</li>
          <li>  git fetch origin $BEFORE &gt;/dev/null 2&gt;/dev/null</li>
          <li>  git diff -z --name-only FETCH_HEAD..HEAD</li>
          <li>else</li>
          <li>  INPUT_ONLY_CHECK_CHANGED_FILES=</li>
          <li>  git 'ls-files' -z 2&gt; /dev/null</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
(I’m not sure you’d really want to try to understand that black magic.)</p>
<p>But, the code is there to support this feature:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/action.yml#L130-L131">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/action.yml#L130-L131" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/action.yml#L130-L131" target="_blank" rel="noopener nofollow ugc">check-spelling/check-spelling/blob/d3e67e88f93da3534317e3cafc06a387bd7daed3/action.yml#L130-L131</a></h4>



  <pre class="onebox">    <code class="lang-yml">
      <ol class="start lines" start="130" style="counter-reset: li-counter 129 ;">
          <li>only_check_changed_files:</li>
          <li>  description: 'If set, only check files changed since the last push'</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<hr>
<p>Essentially you’ll be writing a workflow that will run, and it will use run some action to decide which files have changed since the last push, and then the later steps will be skipped.</p>
<p>That said…</p>
<p>The naive solution is “don’t mix doc changes with code changes unless you want to have testing recheck”. That is a valid approach, although it means you now get 2 PRs, one to add a feature and one to document the feature.</p>
